<?xml version="1.0"?>
<robot name="sensor_r200" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:include filename="$(find coslam)/gazebo/models/rgbd_camera/common_properties.urdf.xacro"/>

  <xacro:property name="r200_cam_rgb_px" value="0.005"/>
  <xacro:property name="r200_cam_rgb_py" value="0.018"/>
  <xacro:property name="r200_cam_rgb_pz" value="0.013"/>
  <xacro:property name="r200_cam_depth_offset" value="0.00"/>

  <xacro:arg name="offset_pz" default="5"/>
  <xacro:arg name="offset_r" default="0"/>
  <xacro:arg name="offset_p" default="0"/>
  <xacro:arg name="offset_y" default="0"/>
  <xacro:arg name="namespace" default="rgb0"/>
  <xacro:arg name="base_x" default="0"/>
  <xacro:arg name="base_y" default="0"/>
  <xacro:arg name="b_yaw" default="0"/>
  

  <link name="camera_base"/>

  <joint name="camera_base_joint" type="fixed">
    <origin xyz="$(arg base_x) $(arg base_y) 0" rpy="0 0 $(arg b_yaw)"/>
    <parent link="camera_base"/>
    <child link="camera_base_link" />
  </joint>
  <link name="camera_base_link">
    <inertial>
      <mass value="1000" />
      <inertia ixx="0.0" ixy="0.0" ixz="0.0"
               iyy="0.0" iyz="0.0"
               izz="0.0" />
    </inertial>
    <collision name='collision'>
      <geometry>
        <box size="0.1 0.1 0.01"/>
      </geometry>
    </collision>
    <visual name='visual'>
      <geometry>
        <box size="0.1 0.1 0.01"/>
      </geometry>
      <material name="Gazebo/Grey" uri="file://media/materials/scripts/gazebo.material"/>
    </visual>
  </link>

  <joint name="camera_joint" type="fixed">
    <origin xyz="0 0 $(arg offset_pz)" rpy="$(arg offset_r) $(arg offset_p) $(arg offset_y)"/>
    <parent link="camera_base_link"/>
    <child link="camera_link" />
  </joint>
  <link name="camera_link">
    <visual>
     <origin xyz="0 0 0" rpy="${M_PI/2} 0 ${M_PI/2}"/>
      <geometry>
       <mesh filename="$(find coslam)/gazebo/models/rgbd_camera/meshes/sensors/r200.dae" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.130 0.567 0.118"/>
      </geometry>
    </collision>

   <inertial>
      <mass value="1" />
      <origin xyz="0 0 0" />
      <inertia ixx="0.003881243" ixy="0.0" ixz="0.0"
               iyy="0.000498940" iyz="0.0"
               izz="0.003879257" />
    </inertial>
  </link>


  <joint name="camera_rgb_joint" type="fixed">
    <origin xyz="${r200_cam_rgb_px} ${r200_cam_rgb_py} ${r200_cam_rgb_pz}" rpy="0 0 0"/>
    <parent link="camera_link"/>
    <child link="camera_optical_frame" />
  </joint>
  <!-- <link name="camera_rgb_frame"/>

  <joint name="camera_rgb_optical_joint" type="fixed">
    <origin xyz="0 0 0" rpy="${-M_PI/2} 0 ${-M_PI/2}" />
    <parent link="camera_rgb_frame" />
    <child link="camera_rgb_optical_frame" />
  </joint>
  <link name="camera_rgb_optical_frame"/>


  <joint name="camera_depth_joint" type="fixed">
      <origin xyz="0 ${r200_cam_depth_offset} 0" rpy="0 0 0" />
      <parent link="camera_rgb_frame" />
      <child link="camera_depth_frame" />
  </joint>
  <link name="camera_depth_frame"/>

  <joint name="camera_depth_optical_joint" type="fixed">
      <origin xyz="0 0 0" rpy="${-M_PI/2} 0 ${-M_PI/2}" />
      <parent link="camera_depth_frame" />
      <child link="camera_optical_frame" />
  </joint> -->
  <link name="camera_optical_frame"/>
  
	
  <gazebo reference="camera_optical_frame">  
    <sensor type="depth" name="camera">
      <always_on>true</always_on>
      <update_rate>60.0</update_rate>
      <camera>
        <horizontal_fov>1.22173</horizontal_fov>
        <image>
          <format>B8G8R8</format>
          <width>640</width>
          <height>480</height>
        </image>
        <clip>
          <near>0.1</near>
          <far>20.0</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <mean>0</mean>
          <stddev>0.2</stddev>
        </noise>
      </camera>
      <plugin name="kinect_camera_controller" filename="libgazebo_ros_openni_kinect.so">
        <cameraName>camera</cameraName>
        <robotNamespace>/$(arg namespace)</robotNamespace>
        <alwaysOn>true</alwaysOn>
        <updateRate>60</updateRate>
        <imageTopicName>image_raw</imageTopicName>
        <depthImageTopicName>depth/image_raw</depthImageTopicName>
        <pointCloudTopicName>depth/points</pointCloudTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <depthImageCameraInfoTopicName>depth/camera_info</depthImageCameraInfoTopicName>
        <frameName>camera_optical_frame</frameName>
        <baseline>0.1</baseline>
        <distortion_k1>0.0</distortion_k1>
        <distortion_k2>0.0</distortion_k2>
        <distortion_k3>0.0</distortion_k3>
        <distortion_t1>0.0</distortion_t1>
        <distortion_t2>0.0</distortion_t2>
        <pointCloudCutoff>0.4</pointCloudCutoff>
      </plugin>
    </sensor>
  </gazebo>

  <!-- <this one cannot be used with other velocity control> -->
  <!-- <gazebo>
    <plugin name="object_controller" filename="libgazebo_ros_planar_move.so">
      <commandTopic>rgbd_cmd</commandTopic>
      <odometryTopic>odom</odometryTopic>
      <odometryFrame>$(arg namespace)/camera_base</odometryFrame>
      <odometryRate>20.0</odometryRate>
      <robotBaseFrame>odom</robotBaseFrame>
    </plugin>
  </gazebo>   -->

  <!-- <gazebo>
    <plugin name="joint_state_publisher" filename="libgazebo_ros_joint_state_publisher.so">
       <jointName>your_joint</jointName>
       <updateRate>50.0</updateRate>
       <alwaysOn>true</alwaysOn>
    </plugin>
  </gazebo> -->
</robot>
